### ISSUE TITLE
When using the embedded dns server, running containers always use the original nameservers

### ISSUE THREAD

**Issue by brandonstevens** [docker/docker#25452]
**Output of `docker version`:**

```
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 4
 Running: 2
 Paused: 0
 Stopped: 2
Images: 2
Server Version: 1.12.0
Storage Driver: devicemapper
 Pool Name: docker-202:1-335555351-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 546.5 MB
 Data Space Total: 107.4 GB
 Data Space Available: 105.3 GB
 Metadata Space Used: 1.507 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.146 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2016-06-09)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.10.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.389 GiB
Name: ip-10-70-50-244
ID: T66J:FYY2:IGT6:HVBU:DMJ3:6MMR:XE4P:63JI:VMJJ:3M2Q:5SXN:JVNW
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Tested using an EC2 instance in AWS.

**Steps to reproduce the issue:**
1. Create a user-defined network (`docker network create dnsissue`)
2. Break resolver conf file on host (`vim /etc/resolv.conf` - change nameserver to something unavailable)
3. Create container on user-defined network (`docker run -d --net=dnsissue --name test nginx`)
4. Fix resolver conf file on host by undoing change made in step 2

**Describe the results you received:**

DNS continues to be broken inside the container. If the container is restarted, DNS is fixed. Also, if a new container is launched on the same network, DNS will work.

**Describe the results you expected:**

The embedded DNS server should update if the hosts resolvers change.

**Additional information you deem important (e.g. issue happens only occasionally):**

It would be helpful to understand how the embedded DNS server works and how two containers on the same network can have different behavior resolving names.

**Comment by brandonstevens** [comment]
I was able to reproduce this on the `master` branch:

```
root@aade163ec0e8:/go/src/github.com/docker/docker# cat /etc/resolv.conf
search local
nameserver 192.168.65.1
nameserver 192.168.65.3`
root@aade163ec0e8:/go/src/github.com/docker/docker# docker exec -it test cat /etc/resolv.conf
search local
nameserver 127.0.0.11
options ndots:0
root@aade163ec0e8:/go/src/github.com/docker/docker# docker exec -it test dig @127.0.0.11 google.com

; <<>> DiG 9.10.3-P4-Ubuntu <<>> @127.0.0.11 google.com
; (1 server found)
;; global options: +cmd
;; connection timed out; no servers could be reached
root@aade163ec0e8:/go/src/github.com/docker/docker# docker exec -it test dig @192.168.65.1 google.com

; <<>> DiG 9.10.3-P4-Ubuntu <<>> @192.168.65.1 google.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 9755
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;google.com.   			IN     	A

;; ANSWER SECTION:
google.com.    		299    	IN     	A      	172.217.0.238

;; Query time: 29 msec
;; SERVER: 192.168.65.1#53(192.168.65.1)
;; WHEN: Sat Aug 20 14:53:08 UTC 2016
;; MSG SIZE  rcvd: 55
```

Debug logs:

```
time="2016-08-20T14:52:53.255728464Z" level=debug msg="Name To resolve: google.com."
time="2016-08-20T14:52:53.256732459Z" level=debug msg="Query google.com.[1] from 172.19.0.2:44826, forwarding to udp:192.168.66.1"
time="2016-08-20T14:52:56.256719189Z" level=debug msg="Read from DNS server failed, read udp 172.19.0.2:48691->192.168.66.3:53: i/o timeout"
```

docker info:

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 1.13.0-dev
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 17
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host null bridge overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.4.15-moby
Operating System: Debian GNU/Linux 8 (jessie) (containerized)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.788 GiB
Name: aade163ec0e8
ID: EZWH:2IDD:2MZW:MDDD:ICMQ:MKNQ:B3R7:GY25:3LXL:IOEQ:TM4F:37I2
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 19
 Goroutines: 32
 System Time: 2016-08-20T14:56:35.370289645Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by sanimej** [comment]
@brandonstevens This is by design. Nameservers from host's resolv.conf at the time of container creation are used as external resolvers for the container. But changes to the file after container creation are not picked up. If you expect changes to host's resolv.conf often (unlikely scenario in most deployments) you can use the `--dns` option to pass the external resolvers explicitly during container creation.

Note that this behavior is not specific to embedded DNS server. Before embedded DNS server was introduced the external resolvers would be updated directly in container's resolv.conf. But the behavior is the same.

I am closing this issue since its the expected behavior.

**Comment by brandonstevens** [comment]
Ahh, you're right. I'm not sure what I saw earlier, but it was likely that the container I was looking at was restarted after the nameservers were updated. We pushed out new nameservers using DHCP options and noticed that a mix of containers were using old and new nameservers.

I think we also misunderstood this bit of documentation on https://docs.docker.com/engine/userguide/networking/default_network/configure-dns/.

> This arrangement allows Docker to do clever things like keep resolv.conf up to date across all containers when the host machine receives new configuration over DHCP later.

Nonetheless, it's good to know that running containers won't update their nameservers and we can plan accordingly when we need to make changes to our nameservers.
